opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -1219,14 +1227,13 @@ namespace Opm
         std::fill(this->ipr_b_.begin(), this->ipr_b_.end(), 0.);
 
         const int nseg = this->numberOfSegments();
-        double seg_bhp_press_diff = 0;
-        double ref_depth = this->ref_depth_;
+        const double ref_depth = this->ref_depth_;
         for (int seg = 0; seg < nseg; ++seg) {
             // calculating the perforation rate for each perforation that belongs to this segment
             const double segment_depth = this->segmentSet()[seg].depth();
-            const double dp = wellhelpers::computeHydrostaticCorrection(ref_depth, segment_depth, this->segment_densities_[seg].value(), this->gravity_);
-            ref_depth = segment_depth;
-            seg_bhp_press_diff += dp;
+            const int outlet_segment_index = this->segmentNumberToIndex(this->segmentSet()[seg].outletSegment());
+            const double segment_depth_outlet = seg == 0? ref_depth : this->segmentSet()[outlet_segment_index].depth();
+            const double dp = wellhelpers::computeHydrostaticCorrection(segment_depth_outlet, segment_depth, this->segment_densities_[seg].value(), this->gravity_);
```

Comment:
At first I thought this was a nice bugfix, but now I think it maybe introduces a bug. The old code accumulates the dp values to form the hydrostatic pressure difference between the bhp reference point and the segment. I think it is mostly correct, except when encountering a fork in the well, there the new code correctly looks at the outlet segment rather than the previous segment.

However, when used further down on line 1264, the dp is used, rather than their sum accumulated up the segments. I believe that is wrong, that the pressure difference should be between segment and bhp ref, not between segment and outlet segment.

I think building a vector of segment-to-bhpref pressure differences is necessary, because to compute a lower one you need to add the difference of the outlet segment to your own. This should be a separate for loop to simplify reading of the code.

Reply:
Seems reasonable. In `3D_MSW` all branches branch out from the top cell. i.e. the issue you point to never surface. I will fix asap.

Reply:
I think it will surface then too, unless the number of segments in each branch is only one?

Reply:
Yes I think that looks good. This will change the results, maybe significantly, so I fear a new round of tests might be required?